connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -307,15 +307,25 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
     }
 
     private void processTargetStateChanges(Set<String> connectorTargetStateChanges) {
-        if (!connectorTargetStateChanges.isEmpty()) {
-            for (String connector : connectorTargetStateChanges) {
-                if (worker.connectorNames().contains(connector)) {
-                    TargetState targetState = configState.targetState(connector);
-                    worker.setTargetState(connector, targetState);
-                    if (targetState == TargetState.STARTED)
-                        reconfigureConnectorTasksWithRetry(connector);
-                }
+        for (String connector : connectorTargetStateChanges) {
+            TargetState targetState = configState.targetState(connector);
+            if (targetState == null) {
+                // The target state was deleted. This should follow only after deletion of the
+                // connector's configuration itself, so raise an error if otherwise.
+                if (configState.connectors().contains(connector))
+                    throw new ConnectException("Unexpected removal of target state for connector " + connector);
```

Comment:
Your explanation here makes sense -- when recreated the config backing stores set the target state (and do so only on creation since there are checks for whether the connector already exists), so it looks like your original implementation is fine -- you really shouldn't be able to get into that state. I'm not sure if there are still some cases where we can get in that state in the case of failures (due to the configs being written separately), but it looks like the current logic is fine.

So now I think you've convinced me that your current patch is fine, but your last comment suggests there's something else you wanted to fix? Or just that you wanted to keep the exception in there?